File: leet-code/454.四数相加-ii.js
/*
 * @lc app=leetcode.cn id=454 lang=javascript
 *
 * [454] 四数相加 II
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number[]} nums3
 * @param {number[]} nums4
 * @return {number}
 */
var fourSumCount = function (nums1, nums2, nums3, nums4) {
  // 转为两数之和
  let count = 0
  const totalMap = new Map()
  for (const item1 of nums1) {
    for (const item2 of nums2) {
      const total = item1 + item2
      let value = totalMap.has(total) ? totalMap.get(total) : 0
      value++
      totalMap.set(total, value)
    }
  }

  for (const item3 of nums3) {
    for (const item4 of nums4) {
      const target = 0 - (item3 + item4)
      if (totalMap.has(target)) {
        count += totalMap.get(target)
      }
    }
  }
  return count
};
// @lc code=end

File: leet-code-rs/src/solutions/454_四数相加-ii.rs
/*
 * @lc app=leetcode.cn id=454 lang=rust
 *
 * [454] 四数相加 II
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn four_sum_count(
        nums1: Vec<i32>,
        nums2: Vec<i32>,
        nums3: Vec<i32>,
        nums4: Vec<i32>,
    ) -> i32 {
        let mut cnt = HashMap::new();
        for n1 in nums1 {
            for n2 in &nums2 {
                *cnt.entry(n1 + n2).or_insert(0) += 1;
            }
        }
        let mut ans = 0;
        for n3 in nums3 {
            for n4 in &nums4 {
                let sum = -(n3 + n4);
                ans += cnt.get(&sum).unwrap_or(&0);
            }
        }

        return ans;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_four_sum_count() {
        assert_eq!(
            Solution::four_sum_count(vec![1, 2], vec![-2, -1], vec![-1, 2], vec![0, 2]),
            2
        );
        assert_eq!(
            Solution::four_sum_count(vec![0], vec![0], vec![0], vec![0]),
            1
        );
        assert!(true)
    }
}
